### DISCUSSION TITLE
slope_per_mill calculation

### DISCUSSION THREAD

**Discussion by spch-GL** [QGEP/QGEP#766] [Q&A] [answered]
I need explanations about the slope calculation formula which is:
```
CASE
            WHEN rp_from.level > 0::numeric AND rp_to.level > 0::numeric THEN round((rp_from.level - rp_to.level) / re.length_effective * 1000::numeric, 1)
            ELSE NULL::numeric
END

```
I do not understand the use of **re.length_effective** in this formula because it is not necessarily a 2D value

**Comment by sjib** [comment]
Yes it includes the 3D information from the reach points and calculates the 3D length:
Definition for reach.length_effective: "Tatsächliche schräge Länge (d.h. nicht in horizontale Ebene projiziert)  inklusive Kanalkrümmungen"
Translated with deepl: "Actual oblique length (i.e., not projected in horizontal plane) including channel curvatures."

**Comment by spch-GL** [comment]
Yes, I understand the calculation of length_effective. But in the slope_per_mill's formula, the length used must be 2D, otherwise the result is not the expected one.

**Reply by sjib** [reply]
@spch-GL Can you add a link to the formula you have in mind?

**Reply by sjib** [reply]
https://github.com/QGEP/datamodel/blob/dcfc70131ad4fa90b99e690b0b59a426fa715f22/view/vw_qgep_reach.py#L45

**Reply by sjib** [reply]
`        CASE `
`          WHEN rp_from.level > 0 AND rp_to.level > 0 THEN round((rp_from.level - rp_to.level)/re.length_effective*1000,1) `
`          ELSE NULL `
`        END AS _slope_per_mill`

rp_from.level - rp_to.level is a difference between two levels -> number. And re.length_effective is also a number. So I do not see why 2D/3D could conflict here.

**Comment by spch-GL** [comment]
I don't know how to link it, but it's the one in the top post: `(rp_from.level - rp_to.level) / re.length_effective`

**Reply by sjib** [reply]
I still do not understand your problem, may be  https://github.com/QGEP/QGEP/issues/315 helps you to understand.

**Comment by spch-GL** [comment]
What I mean is that the formula should not use re.length_effective but it's projection in horizontal plane i.e. ST_Length(re.progression_geometry)

**Reply by spch-GL** [reply]
![Slope](https://user-images.githubusercontent.com/26377323/199309882-1ad29823-66e7-4340-b79e-cd5649f0255b.jpg)
For instance, in this example, the rounded slope is 200 per mill and not 196 per mill

**Reply by sjib** [reply]
Thanks for the documentation. 
In hydraulic simulations this will only be a relevant difference when you have very steep slopes, e.g. in a mountain village.
Related issues to consider also and understand the history: 
https://github.com/QGEP/QGEP/issues/496
https://github.com/QGEP/QGEP/issues/315
https://github.com/QGEP/QGEP/issues/426

@spch-GL Can you formulate a proposal on what you would like to have changed in the code, considering that the reach.progression can be 2D (when z-coordinates are missing or set to NULL or 0.00) or 3D

**Reply by spch-GL** [reply]
My proposal would be something like this (not sure that **progression_geometry** can be used like this in that view) :
```
        CASE
            WHEN rp_from.level > 0::numeric AND rp_to.level > 0::numeric THEN round((rp_from.level - rp_to.level) / ST_Length(progression_geometry) * 1000::numeric, 1)
            ELSE NULL::numeric
        END AS _slope_per_mill,
```

**Reply by sjib** [reply]
Is ST_Length always a projection of the length of a geomtry independant whether it is 2D or 3D?

**Reply by spch-GL** [reply]
According to the doc (https://postgis.net/docs/ST_Length.html) yes

**Reply by spch-GL** [reply]
By the way, the difference is noticeable from 46.5 per mill, wich is not very steep

**Comment by spch-GL** [comment, marked as answer]
Now I think it's a bug. What do you think @3nids @m-kuhn @varrieta ?

**Reply by sjib** [reply]
Issue opened: https://github.com/QGEP/QGEP/issues/776
Fixed with https://github.com/QGEP/datamodel/pull/195
